### PR TITLE
Add solution for TASK-4

### PR DESCRIPTION
**Pull request by perkeje**

### CODE REVIEW

File: src/main/java/net/barrage/school/java/ecatalog/app/services/ProductServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+@Service
+public class ProductServiceImpl implements ProductService {
+
+    @Autowired
+    private ProductRepository productRepository;
+
+
+    @Override
+    @Cacheable("products")
```
Comment (AntonRomankovSetronica) [inline src/main/java/net/barrage/school/java/ecatalog/app/services/ProductServiceImpl.java]: Seems like u dont need to cache it anymore

File: src/main/java/net/barrage/school/java/ecatalog/app/services/ProductServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    public List<Product> listProducts() {
+        return StreamSupport.stream(productRepository.findAll().spliterator(), false)
```
Comment (AntonRomankovSetronica) [inline src/main/java/net/barrage/school/java/ecatalog/app/services/ProductServiceImpl.java]: Simple way - `Lists.newArrayList(productRepository.findAll().iterator())`

File: src/main/java/net/barrage/school/java/ecatalog/web/ProductController.java [hunk 3]
```diff
@@ -29,6 +36,34 @@ public List<Product> listProducts() {
         return products;
     }
 
+    @PostMapping
+    public Product createProduct(@Valid @RequestBody Product product) {
+        var createdProduct = productService.saveProduct(product);
```
Comment (AntonRomankovSetronica) [inline src/main/java/net/barrage/school/java/ecatalog/web/ProductController.java:41]: Err.. What merchant this product will belong?

File: src/main/java/net/barrage/school/java/ecatalog/app/services/ProductSyncServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+            if(source.getMerchantName().equalsIgnoreCase(name)) {
+                syncWithProductSource(source);
+            }
+        }
+    }
+
+    @Transactional
+    private void syncWithProductSource(ProductSource source) {
+        Merchant merchant = merchantRepository.findByName(source.getMerchantName());
+        if(merchant  != null) merchant.getProducts().clear();
```
Comment (AntonRomankovSetronica) [inline src/main/java/net/barrage/school/java/ecatalog/app/services/ProductSyncServiceImpl.java]: - Pls avoid using `if` without `{}`. It's kind of tradition)
- Have u tried this `merchant.getProducts().clear()`? Will try myself, but it looks wrong) Usually u can delete hibernate entity only by calling `delete*` explicitly.

File: src/main/java/net/barrage/school/java/ecatalog/web/MerchantController.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    @GetMapping("/{merchant_id}")
+    public Merchant getMerchantById(@PathVariable("merchant_id") Long merchantId) {
+        var merchant = merchantService.getMerchantById(merchantId);
+        log.trace("getMerchantById -> {}", merchant);
+        return merchant;
+    }
+
+    @GetMapping("/{merchant_id}/products")
+    public Set<Product> listProductsByMerchantId(@PathVariable("merchant_id") Long merchantId) {
+        var products = merchantService.listProductsByMerchant(merchantId);
```
Comment (AntonRomankovSetronica) [inline src/main/java/net/barrage/school/java/ecatalog/web/MerchantController.java]: imho `listProductsByMerchant` should be a part of ProductService API, not MerchantService

File: src/main/java/net/barrage/school/java/ecatalog/web/ProductController.java (hunk 3 above)
Comment (perkeje) [reply src/main/java/net/barrage/school/java/ecatalog/web/ProductController.java:41]: I implemented it the way where client will have to put merchant information in request body

File: build.gradle [hunk 6]
```diff
@@ -24,6 +24,7 @@ dependencies {
     implementation 'org.apache.commons:commons-lang3'
     implementation 'org.apache.poi:poi:5.2.2'
     implementation 'org.apache.poi:poi-ooxml:5.2.2'
+    implementation 'org.springframework.boot:spring-boot-starter-validation'
```
Comment (AntonRomankovSetronica) [inline build.gradle:27]: Nice)

File: src/main/java/net/barrage/school/java/ecatalog/app/services/ProductSyncServiceImpl.java (hunk 4 above)
Comment (perkeje) [reply src/main/java/net/barrage/school/java/ecatalog/app/services/ProductSyncServiceImpl.java]: I tried it and it worked, but i can change it to delete explicitly if it is more clear

File: src/main/java/net/barrage/school/java/ecatalog/app/services/ProductSyncService.java [hunk 7]
```diff
@@ -0,0 +1,8 @@
+package net.barrage.school.java.ecatalog.app.services;
+
+import jakarta.transaction.Transactional;
+
+public interface ProductSyncService {
+    void seedDatabase();
```
Comment (AntonRomankovSetronica) [inline src/main/java/net/barrage/school/java/ecatalog/app/services/ProductSyncService.java]: Why do u need this api? It's insecure to use it, and even to have)

Comment (perkeje) [reply src/main/java/net/barrage/school/java/ecatalog/app/services/ProductSyncService.java]: I wanted to have function that will get all merchants from config and seed the database, but I will delete it now.

File: src/main/java/net/barrage/school/java/ecatalog/web/ProductController.java (hunk 3 above)
Comment (AntonRomankovSetronica) [reply src/main/java/net/barrage/school/java/ecatalog/web/ProductController.java:41]: Oh ok, sorry! By bad, was totally sure I asked for following API, but was wrong)

* POST `/e-catalog/api/v1/merchants/{merchantId}/products` 

There are 2 approaches to solving belongings on creation. One is what u have implemented, another is through extending merchant endpoints like in my example here. Both are ok.

File: src/main/java/net/barrage/school/java/ecatalog/app/services/ProductSyncServiceImpl.java (hunk 4 above)
Comment (AntonRomankovSetronica) [reply src/main/java/net/barrage/school/java/ecatalog/app/services/ProductSyncServiceImpl.java]: Strange, but it works) Okay)

File: src/main/java/net/barrage/school/java/ecatalog/app/services/ProductSyncServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+    @Transactional
+    public void syncWithMerchantName(String name) {
+        for (ProductSource source : productSources) {
+            if(source.getMerchantName().equalsIgnoreCase(name)) {
+                syncWithProductSource(source);
+            }
+        }
+    }
+
+    @Transactional
```
Comment (AntonRomankovSetronica) [inline src/main/java/net/barrage/school/java/ecatalog/app/services/ProductSyncServiceImpl.java]: U have a mess with `Transactional` here. Only public methods can be annotated like that. This one will never work!

**Review by AntonRomankovSetronica — APPROVED** [review]

**Pull request merged**